### PR TITLE
Add ability to generate access tokens

### PR DESCRIPTION
**Pull request by imor**
This PR adds a new feature to allow users to generate access tokens which will be used by the `dbdev login` command (in a separate PR).

In the UI the following changes have been made:

1. In the navbar drop-down menu a new entry `Access Tokens` has been added.
image

2. Clicking `Access Tokens` will take the user to the access tokens page (initially empty). 
image

3. Clicking `New Token` will open the new token form.
image

4. Clicking `Cancel` will take the user back to the access tokens page. Filling a token name and clicking `Create Token` will take the user to the newly created token. On this page the user can copy the token to be pasted into the terminal when `dbdev login` command will be issued. This will also be the only opportunity user will get to copy the token, this is because tokens in the db are stored hashed and can't be shown in plaintext again. 
image

5. Clicking `Close` will take the user back to the access tokens page. Here user can click `Revoke` to delete/revoke a token. 
image
 
In the db tokens are stored in `app.access_tokens` table. RLS policies are setup so that users can only create, read, or delete their own tokens. Tokens are stored salted and hashed in the db. The token returned to the user in the `app.new_access_token` is in plaintext. This token also contains the user id prefixed to the token. This is done so that a single token can act as both a "user id and password" when the `app.redeem_access_token` function is given this token. The `app.redeem_access_token` function validates the token presented to it and issues a jwt with an `authenticated` claim. This jwt can then be used by the client (`dbdev` in our case) to make api calls.

### CODE REVIEW

File: website/data/access-tokens/create-access-token.ts
Comment (alaister) [inline website/data/access-tokens/create-access-token.ts:46]: Is there any reason you're using `resetQueries` here instead of just invalidating the `['access-tokens']` query?
For example:
```typescript
queryClient.invalidateQueries(['access-tokens'])
```

File: website/data/access-tokens/delete-access-token.ts
Comment (alaister) [inline website/data/access-tokens/delete-access-token.ts:46]: Same here

File: website/pages/[handle]/_/access-tokens.tsx
Comment (alaister) [inline website/pages/[handle]/_/access-tokens.tsx:117]: react-query is smart enough to make all of these options mutually exclusive so we can avoid the ternaries here :)
For example:
```tsx
<>
  {isLoading && <p>Loading...</p>}
  {isSuccess && ...}
  {isError && <p className="text-red-500">Error loading access tokens</p>}
</>
```

**Review by alaister — COMMENTED** [review]
Looking 🔥🔥🔥
Mainly reviewed the FE code – just a couple of small comments 🙂

**Review by olirice — COMMENTED** [review]
the SQL looks great!

lets wait for pgsodium review from michel and auth flow review from stojan before approving

I ran into some minor UI stuff during testing (video below)

https://github.com/supabase/dbdev/assets/12958657/632308e3-c327-476f-9f3f-73bbd34f1bba

The 3 I noticed were:
- `Loading...` not visible in dark mode
- clicking `Create Token` did not redirect or have any visible effect
- `Error loading access tokens` after creating an access token

**Review by michelp — CHANGES_REQUESTED** [review]

File: supabase/migrations/20230817110845_access_token.sql
Comment (kangmingtay) [inline supabase/migrations/20230817110845_access_token.sql:3]: i think we should cascade deletes here since an access token shouldn't exist if a user is deleted
```suggestion
    user_id uuid not null references auth.users(id) on delete cascade,
```

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:41]: May be easier for maintenance to have one permissive policy `for all` that enforces the `auth.uid() = user_id` binding, and then add restrictive policies if something needs stopping (like deletes or the such).

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [reply supabase/migrations/20230817110845_access_token.sql:54]: There's no need to do this. Password hashing functions are only useful and worth the CPU cycles if the thing they're hashing is a password -- i.e. a string of text not-very-randomly chosen from a not-very-big dictionary (e.g. chosen by a person, using common words or dates).

You are very safe to just do SHA256 or any high-quality hashing function here because you're literally using the highest quality randomness source available to generate the token in the first place. We do this on the Supabase platform extensively.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:62]: This will quite literally never happen, so I would omit it and panic when you see the insert violation, as the token being inserted will never repeat while this universe exists.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:83]: Ditto, you don't need to do this. Just hash it with SHA256.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:84]: It's often useful to encode some identifying information in the token's prefix. For example: `dbdev_<token>`.

Also what you may want to store in the database is the token's prefix (first 4-5 chars) to help people identify it quickly. It may also be a bit more compact to encode as Base64-URL or something.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:148]: Again, no need to do password hash verification, which is also VERY CPU intensive. When you switch to SHA256, make sure you don't just do `=` as that's not constant-time comparison. I need to look into how this can be achieved in Postgres easily as I'm not finding it but @michelp this may be a good API to add to pgsodium.

In general though, this is really a fringe security benefit for this use case.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [inline supabase/migrations/20230817110845_access_token.sql:171]: Always useful to include the `iss` claim that will identify who issued the JWT.

**Review by hf — COMMENTED** [review]
Looks good, though the password hashing is really not necessary for this use case and can actually severely tank performance of your database if someone puts their mind to it (which isn't that difficult). So I would just use SHA256 in the name of security (CIA model -- Availability is impacted by using password hashing when you don't need it without effective rate limiting or features providing for fairness).

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [reply supabase/migrations/20230817110845_access_token.sql:171]: Also, you may want to identify the key that signed it with the `kid` claim in the JWT's header. This can be useful to easily identity and rotate the key (in Supabase PostgREST we're just about to add support for this).

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [reply supabase/migrations/20230817110845_access_token.sql:148]: Oh also I suppose libsodium uses Argon2 which is _incredibly_ CPU and memory intensive (depending on the settings). So this is really an easy way to crash the database -- just start generating / verifying tokens all the time.

File: supabase/migrations/20230817110845_access_token.sql
Comment (hf) [reply supabase/migrations/20230817110845_access_token.sql:62]: Ah sorry, it's about the name. It's fine, then though I don't see the point in enforcing name uniqueness -- it's just for the UX. By making the token's prefix available in the database, debugging will be easier for people.

File: website/data/access-tokens/create-access-token.ts
Comment (imor) [reply website/data/access-tokens/create-access-token.ts]: It was a copy paste mistake. I've changed it to invalidate only the 'access-tokens' query.

File: website/data/access-tokens/delete-access-token.ts
Comment (imor) [reply website/data/access-tokens/delete-access-token.ts]: Fixed.

File: website/pages/[handle]/_/access-tokens.tsx
Comment (imor) [reply website/pages/[handle]/_/access-tokens.tsx]: Didn't know that, thank you. Updated.

**Comment by imor** [comment]
@olirice I have fixed the invisible `Loading...` text in dark mode. Regarding the other two problems, it looks like as if the API is not working. Did you run the migrations or reset the db?

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql:53]: Looks like `authenticated` user doesn't have permission to run `pgsodium.ramdombytes_buf`:

```sql
postgres=> select proname from pg_proc where has_function_privilege('authenticated', oid, 'execute') and proname = 'randombytes_buf' order by proname;
 proname 
---------
(0 rows)
```
Any reason why `authenticated` users can't call it?

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql:171]: I read the signing key with `current_setting('app.settings.jwt_secret', true)`, but how do I get the key id?

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: AFAIK in development the `current_setting('app.settings.jwt_secret', true)` [doesn't work during development](https://github.com/supabase/supabase-js/issues/25#issuecomment-1019935888). But I agree that this can be dangerous to fall back to a hardcoded value, so I have removed it.

Ideally, this setting should be returned in development as well. If not, then if I could read another setting which can tell me the current environment will also work.

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: Thank you @michelp, @hf what you've said make a lot of sense. I'm moving the impl to use SHA256 hash.

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: I have removed the unique constraint on (token_name, user_id).

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: Moved to a SHA256 impl now.

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: Implemented this, tokens now look like this:

image

Still a bit long due to `token_id` being uuids. I'm trying to convert them to an integer which should make the tokens smaller.

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql]: Moved to a SHA256 based impl.

File: supabase/migrations/20230817110845_access_token.sql
Comment (imor) [reply supabase/migrations/20230817110845_access_token.sql:40]: If you look carefully, the topmost policy is for the `public` role and the other two are for `authenticated`. So, adding a `for all` policy and then restricting individual operations will need more policies: two `for all` policies, three policies for restricting `insert`, `update` and, `delete` to `public` and two policies for restricting `select` and `update` to `authenticated`. Which makes seven policies in total.

File: website/pages/[handle]/_/access-tokens.tsx
Comment (alaister) [inline website/pages/[handle]/_/access-tokens.tsx:36]: What's the reason we need to refresh the session before creating an access token?

**Review by alaister — APPROVED** [review]
Approving on behalf of the frontend changes. LGTM!

One small code question but a non-blocker:

File: website/pages/[handle]/_/access-tokens.tsx
Comment (imor) [reply website/pages/[handle]/_/access-tokens.tsx]: Good catch, not needed, removed.

File: website/pages/[handle]/_/access-tokens.tsx
Comment (imor) [inline website/pages/[handle]/_/access-tokens.tsx:83]: Nice!

File: supabase/migrations/20230817110845_access_token.sql
Comment (sweatybridge) [reply supabase/migrations/20230817110845_access_token.sql]: I will release a fix to CLI to set these jwt configs for local development so it matches our hosted platform. https://github.com/supabase/cli/pull/1432

**Review by olirice — APPROVED** [review]

**Pull request merged**
